File: src/main/java/com/flab/infrun/member/infrastructure/MemberVerifierImpl.java
package com.flab.infrun.member.infrastructure;

import static java.util.regex.Pattern.matches;

import com.flab.infrun.member.domain.MemberVerifier;
import com.flab.infrun.member.domain.exception.DuplicatedEmailException;
import com.flab.infrun.member.domain.exception.DuplicatedNicknameException;
import com.flab.infrun.member.domain.exception.InvalidPasswordException;
import com.flab.infrun.member.infrastructure.persistence.MemberJpaRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.util.StringUtils;

@RequiredArgsConstructor
public class MemberVerifierImpl implements MemberVerifier {

    private final MemberJpaRepository memberJpaRepository;

    @Override
    public void verifyNickname(final String nickname) {
        if (!StringUtils.hasText(nickname) || memberJpaRepository.existsByNickname(nickname)) {
            throw new DuplicatedNicknameException();
        }
    }

    @Override
    public void verifyEmail(final String email) {
        if (!StringUtils.hasText(email) || memberJpaRepository.existsByEmail(email)) {
            throw new DuplicatedEmailException();
        }
    }

    @Override
    public void verifyPassword(final String password) {
        if (!StringUtils.hasText(password)
            || !matches("^(?=.*[A-Z])(?=.*\\d)(?=.*[a-z])(?=.*[\\W_])[A-Za-z\\d\\W_]{8,16}$",
            password)
        ) {
            throw new InvalidPasswordException();
        }
    }
}

File: src/main/java/com/flab/infrun/member/infrastructure/MemberVerifierImpl.kt
package com.flab.infrun.member.infrastructure

import com.flab.infrun.member.domain.MemberVerifier
import com.flab.infrun.member.domain.exception.DuplicatedEmailException
import com.flab.infrun.member.domain.exception.DuplicatedNicknameException
import com.flab.infrun.member.domain.exception.InvalidPasswordException
import com.flab.infrun.member.infrastructure.persistence.MemberKtJpaRepository
import lombok.RequiredArgsConstructor
import org.springframework.util.StringUtils
import java.util.regex.Pattern

@RequiredArgsConstructor
class MemberVerifierImpl(
    private val memberJpaRepository: MemberKtJpaRepository
) : MemberVerifier {

    override fun verifyNickname(nickname: String) {
        if (!StringUtils.hasText(nickname) || memberJpaRepository.existsByNickname(nickname)) {
            throw DuplicatedNicknameException()
        }
    }

    override fun verifyEmail(email: String) {
        if (!StringUtils.hasText(email) || memberJpaRepository.existsByEmail(email)) {
            throw DuplicatedEmailException()
        }
    }

    override fun verifyPassword(password: String) {
        if (!StringUtils.hasText(password)
            || !Pattern.matches(
                "^(?=.*[A-Z])(?=.*\\d)(?=.*[a-z])(?=.*[\\W_])[A-Za-z\\d\\W_]{8,16}$",
                password
            )
        ) {
            throw InvalidPasswordException()
        }
    }
}
